controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -92,20 +125,432 @@ func (r *StorageClassClaimReconciler) Reconcile(ctx context.Context, request rec
 	return result, nil
 }
 
+func (r *StorageClassClaimReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&ocsv1alpha1.StorageClassClaim{}, builder.WithPredicates(
+			predicate.GenerationChangedPredicate{},
+		)).
+		Owns(&rookCephv1.CephBlockPool{}).
+		Owns(&rookCephv1.CephFilesystemSubVolumeGroup{}).
+		Owns(&rookCephv1.CephClient{}).
+		Complete(r)
+}
+
 func (r *StorageClassClaimReconciler) reconcileConsumerPhases(ctx context.Context, storagecluster *v1.StorageCluster) (reconcile.Result, error) {
 	log := ctrllog.FromContext(ctx)
 	log.Info("Running StorageClassClaim controller in Consumer Mode")
 	return reconcile.Result{}, nil
 }
 
-func (r *StorageClassClaimReconciler) reconcileProviderPhases(ctx context.Context, storagecluster *v1.StorageCluster) (reconcile.Result, error) {
-	log := ctrllog.FromContext(ctx)
-	log.Info("Running StorageClassClaim controller in Converged/Provider Mode")
+func (r *StorageClassClaimReconciler) reconcileProviderPhases() (reconcile.Result, error) {
+	r.log.Info("Running StorageClassClaim controller in Converged/Provider Mode")
+
+	r.storageClassClaim.Status.Phase = string(v1alpha1.StorageClassClaimStateInitializing)
+
+	storageConsumerList := &ocsv1alpha1.StorageConsumerList{}
+	err := r.Client.List(r.ctx, storageConsumerList, client.InNamespace(r.namespace))
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to list storage consumer list. %v", err)
+	}
+
+	r.storageConsumer = &ocsv1alpha1.StorageConsumer{}
+	r.storageConsumer.Namespace = r.namespace
+	for _, consumer := range storageConsumerList.Items {
+		if r.storageClassClaim.Labels[controllers.ConsumerUUIDLabel] == string(consumer.GetUID()) {
+			r.storageConsumer.Name = consumer.Name
+		}
+	}
+
+	if err = r.get(r.storageConsumer); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if r.storageClassClaim.Spec.Type == "blockpool" {
+		r.cephBlockPool = &rookCephv1.CephBlockPool{}
+		r.cephBlockPool.Name = fmt.Sprintf("cephblockpool-%s", r.storageConsumer.Name)
+		r.cephBlockPool.Namespace = r.namespace
+		r.cephBlockPool.Labels[controllers.StorageConsumerNameLabel] = r.storageConsumer.Name
+
+		r.cephClientProvisioner = &rookCephv1.CephClient{}
+		r.cephClientProvisioner.Name = controllers.GenerateHashForCephClient(r.storageClassClaim.Name, "provisioner")
+		r.cephClientProvisioner.Namespace = r.namespace
+
+		r.cephClientNode = &rookCephv1.CephClient{}
+		r.cephClientNode.Name = controllers.GenerateHashForCephClient(r.storageClassClaim.Name, "node")
+		r.cephClientNode.Namespace = r.namespace
+	} else if r.storageClassClaim.Spec.Type == "sharedfilesystem" {
+		r.cephFilesystemSubVolumeGroup = &rookCephv1.CephFilesystemSubVolumeGroup{}
+		r.cephFilesystemSubVolumeGroup.Name = fmt.Sprintf("cephfilesystemsubvolumegroup-%s", r.storageConsumer.Name)
```

Comment:
This code is identical for both cases, can we move it outside above the if statement?

Reply:
ack